### DISCUSSION TITLE
How does Longhorn handle replicas & split brain ?

### DISCUSSION THREAD

**Discussion by kub3let** [longhorn/longhorn#5486] [Q&A] [answered]
With other hyper converged storage for example Ceph you can not set replicas below a split brain scenario, e. g. if you have 3 nodes you need 3 replicas, since if 1 fails there needs to be a quorum of 2 still available. So for Ceph if you set replicas to 2 and one node fails, storage is set to read only and your cluster goes down.

How does Longhorn handle such scenario ?

If I have a 3 node cluster, is it fine to set replicas to 2 and if one of the nodes fails which has one of the replica will longhorn continue to work ? Will it auto rebalance to the remaining node ? Or will it fail and storage becomes read only ?

Personally I would like to only have 2 replicas, since this increases write speed and reduces wasted storage but it really depends on how Longhorn handles such a scenario.

**Comment by kub3let** [comment, marked as answer]
So I just tried this scenario, killing the replicated node that is not mounted will simply put the volume into a degraded state, but reading/writing to it will still work. Restarting the failed node will automatically resilver the volume and everything is healthy again.

With killing the node which the pod is running on I had some issues, it took 5 minutes until a failover was initiated, but k8s was not able to kill the running pod and it was stuck in terminating state. Due to that the new pod then had the error: Multi-Attach error for volume "pvc-d57395b0-085b-46ad-821c-dd86f2aab468" Volume is already used by pod(s) podname-5b74887575-nhn4s

Waiting for 30 minutes nothing happened. This however has nothing to do with replica count and seems more to be a general issue. Restarting the failed node in that case would successfully terminate the original pod and the backup pod will startup just fine then.

edit: seems to be fixed in k8s 1.27 https://github.com/longhorn/longhorn/issues/3778

**Reply by dberardo-com** [reply]
alright, so what you are saying here is that it is enough to have 3 nodes and 2 replicas distributed on 2 nodes to achieve High availability with longhorn, is this right ?

do you know how to ensure that longhorn does not build those 2 replicas on the same node ? can this be configured as a global setting somewhere or needs to be done on a volume by volume basis ?

and also, do you know if the same statements are true for the read write many volumes ?

**Reply by kub3let** [reply]
alright, so what you are saying here is that it is enough to have 3 nodes and 2 replicas distributed on 2 nodes to achieve High availability with longhorn, is this right ?

Yes technically 2 replicas is enough for redundancy but in a production environment you would want to have >=3

do you know how to ensure that longhorn does not build those 2 replicas on the same node ? can this be configured as a global setting somewhere or needs to be done on a volume by volume basis ?

I never tested that myself but Longhorn is aware of the node the volume is on, I would expect the balancing to be based on node and not volume.

and also, do you know if the same statements are true for the read write many volumes ?

You are best of testing longhorn yourself in a test environment.

**Reply by ghost** [reply]
@dberardo-com 
> do you know how to ensure that longhorn does not build those 2 replicas on the same node ? can this be configured as a global setting somewhere or needs to be done on a volume by volume basis ?

I believe you're looking for data locality

https://longhorn.io/docs/1.7.2/high-availability/data-locality/#data-locality-settings

    strict-local: This option enforces Longhorn keep the only one replica on the same node as the attached volume, and therefore, it offers higher IOPS and lower latency performance.

**Reply by dberardo-com** [reply]
I knew abo data locality, but this with it I would lose high availability, so I never use it. Please let me know if I am wrong here but I believe that if the nodes goes done, then the volume cannot be rebuilt anywhere else as all data live on the node that went down. This is a no go.

I would like to have always at least 2 replicas of a volume on different nodes (best-effort) but the important thing is that new pods are scheduled in affinity with those replicas, i.e. new pods are scheduled on nodes that have replicas on them.

Again, I believe this should be simple to understand, am I overcomplicating the description of the use case?

In essence:
* Volume needs to be highly available (replicas>1)
* Replica rebuild process needs to be reduced to the minimum (i.e. new pods using the volume can only be scheduled on nodes having ready replicas)
* The solution to achieve this needs to scale, I cannot start manually labeling all nodes and pods to use different affinity rules when I have lot of nodes and lot of volumes

**Reply by ghost** [reply]
~~It will create replicas on other nodes but I'm unsure of how "live" that snapshot is. I'm experimenting with this as we speak.~~

This is the pitfall with modern distributed/highly available storage. Say you were to write a file every 1 ms to node1 which is replicated to node2. When/If node1 goes down there will be some sort of data loss depending upon how quick the sync to node 2 was occurring. Even Ceph with its write to all replicas before reply-acknowledge mode will experience data loss in this event.

Unfortunately the best way to answer your own question here is to deploy a simulation of your requirements and test it

EDIT: Nevermind strict-local requires replica `1` be set. My ability to get that working was a fluke.

**Reply by dberardo-com** [reply]
this is the closest answer i found to a solution, but again it is not well scalable: https://github.com/longhorn/longhorn/discussions/9741#discussioncomment-11471331

here i propose using longhorn-system pods as source of affinity, lets see what the answer will be: https://github.com/longhorn/longhorn/discussions/9741#discussioncomment-11481477

**Reply by ghost** [reply]
Thanks for sharing, it sounds like we're trying to get the same thing setup.

I have this scenario setup with Ceph on a few clusters but ceph can be extremely resource intensive. Sometimes I just want to start a quick cluster to experiment and ceph is too heavy-weight for that so I wanted to start exploring with longhorn.

**Reply by dberardo-com** [reply]
does ceph offer a solution for this pod scheduling problem ?

**Reply by ghost** [reply]
Yes with a rook/ceph setup its possible, but you will run into the issue the original post of this discussion mentioned which is minimum amount of nodes/monitors/OSD and your need to possibly not have the volume available ahead of time. In Ceph thinking the volume exists on all possible nodes.

To solve your problem of the scheduling there are a couple of ways to handle it:

- Use a CI/CD solution that has advances scheduling capabilities, which could control all the labeling for you
- a mutating admission controller which can watch for resources being created and modify certain aspects of the definitions/manifests
- you can create the volumes ahead of time, and then watch them to attach and then deploy the workload on the node it was created on
- Use a policy agent with an admission controller to do the mutating resources idea (https://www.openpolicyagent.org/docs/latest/kubernetes-introduction/)


If you use OPA and get creative it can actually do all of the above I just mentioned including integration with CI/CD


EDIT: the term I think you're looking for is generally called HA distributed hyper-converged

**Reply by dberardo-com** [reply]
> . In Ceph thinking the volume exists on all possible nodes.

Sure this is a solution ... but i really want to avoid this :D too wasteful

> Use a CI/CD solution that has advances scheduling capabilities, which could control all the labeling for you

too slow ... i want immediate response, at most 30 seconds to reschedule a pod

> a mutating admission controller which can watch for resources being created and modify certain aspects of the definitions/manifests

this sounds promising ... but i have never dealt with this k8s component before ... do you think this can achieve the scheduling i am looking for, by perhaps interacting with the longhorn API to check where healthy replicas are available ?

> you can create the volumes ahead of time, and then watch them to attach and then deploy the workload on the node it was created on

again, i am looking for incident response here, in case of node faults, so i am not able to predict that in advance

> Use a policy agent with an admission controller to do the mutating resources idea

sounds also promising, is this any different then admission controller? is this a light weight or a "very heavy" solution?

> HA distributed hyper-converged

OMG sounds creepy :D

---

UPDATE: just found out that longhorn is updating the Nodes CRD with a value called scheduledReplicas. Since podaffinity do not work with CRD, i think one could use a custom controller that checks this field and then update the actual Node's labels with the values from this CRD. Then the nodeaffinity rule could work on the pod to be scheduled on those nodes which have replicas on them.
hope this makes sense.

**Reply by ghost** [reply]
I'm over thinking your problem, sorry about possibly misleading in the wrong direction. Those previous suggestions are good for other and more complex needs or if you absolutely cannot find another simpler solution. I'm new to longhorn myself so I'm suggesting more generic approaches.

How are you creating you pvc, pv, and pods? Do you have a simple example to show?


For expert help on scheduling see this page, the slack channel, and community meetings:
https://github.com/kubernetes/community/tree/master/sig-scheduling

**Reply by ghost** [reply]
@dberardo-com  I just read your update, that is great information to know and would work for my case aswell. thanks!

**Reply by dberardo-com** [reply]
May I ask which solution were you thinking of implementing? Is it a custom controller? 

Imif you feel like it we might try to join forces ... I am not a GO developer though ... So I might have to use python koof library for controllers

Let me know

**Reply by ghost** [reply]
> May I ask which solution were you thinking of implementing? Is it a custom controller?
> 
> Imif you feel like it we might try to join forces ... I am not a GO developer though ... So I might have to use python koof library for controllers
> 
> Let me know

So right before you mentioned that I was thinking, whats the absolute simplest way of solving this. And the answer is to just query the k8s API, and apply node labels manually on a cron job say every 5 seconds or something. This would be the "dummy" approach and for my use case (battle testing short-lived cluster configurations) it'll work. 

I would stay away from things that need to be heavily maintained like a custom controller or easily break due to changes in k8s ecosystem. 

Take something [as simple as this example](https://github.com/kubernetes-client/python/blob/master/examples/node_labels.py) to apply the node labels. Then containerize this simple little app and run it as a cronjob in the cluster, or have the container long-lived and run the cron inside the container

Does that make sense?

**Reply by dberardo-com** [reply]
actually i believe that writing a simple custom controller using kopf is the best way to go about it. The controller does not need to poll all the time, as it can  subscribe to internal CRD events and thus it is very efficient. 

Also it is going to be a dumb controller that only acts on CRD events and writes the same label on the same node specified by the CRD, so the code would potentially be even shorter than the one you posted:  https://kopf.readthedocs.io/en/latest/handlers/#state-changing-handlers

**Reply by ghost** [reply]
That is definitely the _right way_ to go about it, don't let my previous message detract from that; however I did not know there were such simple controller abstractions/wrappers available. Kopf does look interesting for sure, a lot easier than what I've worked with and written in the past.

Basic workflow for operator would look something like this:

PV Creation
1. PV created
2. Event caught by Kopf
3. Handler scrapes node/PV relationship from object
4. Use that to create a label on the applicable node

PV Deletion
1. PV is Deleted (or related events)
2. Event caught by Kof
3. Scrape data from the object
4. Get current node labels and delete related PV from list

Along the lines of what you're thinking?

EDIT: Uhh I skipped a step in we'd also need to related PVC to PV and watch for PVC events

Also the "resuming/first startup" would be handled by this https://kopf.readthedocs.io/en/latest/handlers/#resuming-handlers

**Reply by ghost** [reply]
The longhorn kind: Engine schema has a map of the Replica Addresses and is labeled with the volume

```
longhornvolume: pvc-4bc07ed5-843c-4c6e-8746-3c5603d8fe91
```

```
  currentReplicaAddressMap:
    pvc-4bc07ed5-843c-4c6e-8746-3c5603d8fe91-r-6d1d61da: 10.10.10.5:10000
    pvc-4bc07ed5-843c-4c6e-8746-3c5603d8fe91-r-f140ded4: 10.10.10.6:10000
  replicaModeMap:
    pvc-4bc07ed5-843c-4c6e-8746-3c5603d8fe91-r-6d1d61da: RW
    pvc-4bc07ed5-843c-4c6e-8746-3c5603d8fe91-r-f140ded4: RW
```


It looks like there are a few approach possibilities. The `VolumeAttachments` CRD is also interesting, it has node/volume information and the CSI information
